### PR TITLE
Add DM support to the List component

### PR DESCRIPTION
**Pull request by bpauli**
- wrap the current list resource for the teaser items so the content policy for the imageDelegate can take into account
- provide the path to the featuredImage as suffix to the image src attribute
- make the AdpativeImageServlet aware of the feature image suffix
- adapt unit test

relates to #1473

### CODE REVIEW

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/helper/image/AdaptiveImageHelper.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+public class AdaptiveImageHelper {
+
+    public static final String IMAGE_RESOURCE_TYPE = "core/wcm/components/image";
+
+    private AdaptiveImageHelper() {
+
+    }
+
+    @Nullable
+    public static Resource getComponentCandidate(@NotNull String suffix, Resource component) {
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/helper/image/AdaptiveImageHelper.java:47]: It would be nice to add some javadocs.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/helper/image/AdaptiveImageHelper.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+public class AdaptiveImageHelper {
+
+    public static final String IMAGE_RESOURCE_TYPE = "core/wcm/components/image";
+
+    private AdaptiveImageHelper() {
+
+    }
+
+    @Nullable
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/helper/image/AdaptiveImageHelper.java:46]: It would be nice to add some javadocs.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/helper/image/AdaptiveImageHelper.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+                // image coming from featured image
+                Resource featuredImageResource = resourceResolver.getResource(suffixPath);
+                if (featuredImageResource != null && featuredImageResource.isResourceType(IMAGE_RESOURCE_TYPE)) {
+                    componentCandidate = featuredImageResource;
+                }
+            }
+        }
+        return componentCandidate;
+    }
+
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/helper/image/AdaptiveImageHelper.java:74]: It would be nice to add some javadocs.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/helper/image/AdaptiveImageHelper.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            try {
+                requestLastModified = Long.parseLong(ResourceUtil.getName(m.group()));
+            } catch (NumberFormatException e) {
+                // do nothing
+            }
+        }
+        return requestLastModified;
+    }
+
+    @Nullable
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/helper/image/AdaptiveImageHelper.java:93]: It would be nice to add some javadocs.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ImageImpl.java [hunk 5]
```diff
@@ -126,7 +127,8 @@ public class ImageImpl extends AbstractComponentImpl implements Image {
     protected String selector;
     protected String extension;
     protected long lastModifiedDate = 0;
-    protected boolean inTemplate;
+    protected boolean inTemplate = false;
+    protected boolean pageListItem = false;
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ImageImpl.java:131]: I'd rename it to `isPageListItem`

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ImageImpl.java [hunk 6]
```diff
@@ -164,6 +166,10 @@ protected void initModel() {
         fileReference = properties.get(DownloadResource.PN_REFERENCE, String.class);
         alt = properties.get(ImageResource.PN_ALT, String.class);
         title = properties.get(JcrConstants.JCR_TITLE, String.class);
+        featuredImageResourcePath = properties.get(PageListItemImpl.PN_FEATURED_IMAGE_RESOURCE_PATH, String.class);
+        if (StringUtils.isNotEmpty(featuredImageResourcePath)) {
+            pageListItem = true;
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ImageImpl.java:171]: I find it weird that the ImageImpl depends on the List component: the image impl should not know about who is calling it, or in which composite component it is being used.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/PageListItemImpl.java [hunk 7]
```diff
@@ -37,6 +37,8 @@
  */
 public class PageListItemImpl extends AbstractListItemImpl implements ListItem {
 
+    public static final String PN_FEATURED_IMAGE_RESOURCE_PATH = "featuredImageResourcePath";
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/PageListItemImpl.java:40]: I'd define this constant in the image impl.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ImageImpl.java (hunk 6 above)
Comment (bpauli) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ImageImpl.java]: right, in the latest commit https://github.com/adobe/aem-core-wcm-components/pull/2367/commits/cf499ba81168a819b1d2f0a963ae644b529c4967 I made this concept of an external image resource more general which might be used also by other components in the future.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v2/PageListItemImpl.java [hunk 8]
```diff
@@ -118,7 +127,7 @@ public Resource getTeaserResource() {
                         overriddenProperties.put(ImageResource.PN_LINK_URL, this.getPath());
                     }
                 }
-                teaserResource = new CoreResourceWrapper(resourceToBeWrapped, delegateResourceType, hiddenProperties, overriddenProperties);
+                teaserResource = new CoreResourceWrapper(listResource, delegateResourceType, hiddenProperties, overriddenProperties);
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v2/PageListItemImpl.java:130]: Shouldn't it be the `resourceToBeWrapped` instead?

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Image.java [hunk 9]
```diff
@@ -103,6 +103,13 @@ public interface Image extends Component {
      */
     String PN_DISPLAY_POPUP_TITLE = "displayPopupTitle";
 
+    /**
+     * Name of the resource property that will indicate if the image component should render an external image resource.
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Image.java:107]: Isn't it the name of the property holding the path of the external image?

Comment (bpauli) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Image.java:107]: No, it's holding the path to the other image component resource

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v2/PageListItemImpl.java (hunk 8 above)
Comment (bpauli) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v2/PageListItemImpl.java:130]: No, this needs to be the listResource as we want to get the content police of the image delegate component for the same structure element.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Image.java (hunk 9 above)
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/Image.java:107]: I think we mean the same :) . I'd rephrase it as follows:
Name of the property holding the path of the external image resource.

**Review by jckautzmann — APPROVED** [review]
